fastlane/helper/plugin_scores_helper.rb

```diff
@@ -201,6 +222,121 @@ def append_github_data
         end
       end
+
+      class FastlaneActionFileParser
+        attr_accessor :state
+
+        def initialize
+          self.state = []
+        end
+
+        # Parses the file to find all included actions
+        def parse_file(file, debug_state: false)
+          lines = File.read(file).lines
+          actions = []
+
+          # Class state information
+          name = nil
+          description = nil
+          category = nil
+          details = nil
+          # Method state information
+          last_method_name = nil
+          last_string_statement = nil
+
+          lines.each do |line|
+            case self.state.last
+            when nil
+              name = action_name_from_line(line)
+              self.state << :in_class if is_class?(line)
+            when :in_class
+              if (method_name = method_name_from_line(line))
+                last_method_name = method_name
+                self.state << :in_method
+              end
+              if is_end?(line)
+                self.state.pop
+                # Make sure to skip helper classes where the name is nil
+                actions << FastlanePluginAction.new(name: name.fastlane_underscore, category: category, description: description, details: details) unless name.nil?
+                name = category = description = details = nil
+              end
+            when :in_method
+              if (string_statement = string_statement_from_line(line))
+                last_string_statement = string_statement
+              end
+              if is_block?(line)
+                self.state << :in_block
+              end
+              if is_end?(line)
+                category = last_string_statement if last_method_name == "category"
+                description = last_string_statement if last_method_name == "description"
+                details = last_string_statement if last_method_name == "details"
+                last_method_name = last_string_statement = nil
+                self.state.pop
+              end
+            when :in_block
+              if is_block?(line)
+                self.state << :in_block
+              end
+
+              self.state.pop if is_end?(line)
+            end
+            next unless debug_state
+            puts "Current line: #{line}"
+            puts "Full State: #{state}"
+            puts "Current action name: #{name}, category: #{category}, description: \"#{description}\", details: \"#{details}\""
+            puts "Last string statement: #{last_string_statement}, last method name: #{last_method_name}"
+          end
+          actions
+        end
+
+        def is_block?(line)
+          # Matches beginning block statement from the list of keywords
+          # `do` has to be handled separately because it can be preceded by a function call
+          line.match?(/^(?:[\s\w]*=)?\s*(if|unless|case|while|loop|for|until|begin)/) \
+          || line.match?(/^(?:[\s\w]*=)?\s*.*\s+do\s*(?:\|.*\|)?\s*(?:#.*)?$/)
```

Comment:
For the record this regex is not that accurate so might misfire in some cases... I previously used this regex: 

```regex
^(?:[\s\w]*=)?\s*(?:(?:[\w\.]+(?:\[.*\])*)+(?:\(.*\))?)*\s+do\s*(?:\|.*\|)?\s*(?:#.*)?$
```

Which works wonderfully in JS and matches exactly one variable assignment, a variable number of method calls including subscripts, and then a trailing do block. In ruby it hangs the interpreter though so I had to go with the current solution

Reply:
Is there any way to actually require each plugin and use reflection if necessary to find the action classes it implements? That might end up being cleaner or simpler than the regexp approach, but there would be issues with dependencies.

Reply:
I thought about 3 options when coming up with this PR: 

1. Parse the ruby file into an AST using a 3rd party plugin, grab values from AST 
2. `require` the file and grab the metadata we want to show via ruby directly
3. Parse a subset of ruby that includes the information we need and grab values like that

I decided against option 1 because I thought bringing in a 3rd party plugin would introduce additional maintenance burden (if the plugin API changes we need to update that code) and I guessed that the number of lines to parse the file and get the values that we need out of the AST would be comparable in length to option 3

I also decided against option 2 because of security considerations, since we're downloading arbitrary code from the internet. IIRC even if you just require a file top-level code is run, so we would have to build in some safeguards to make sure that no malicious code can be run like that. Even then, the stakes are high and in a worst case scenario if we have a bug in our code an attacker could get access to contributors' computers or CI systems. With options 1 & 3 the worst case scenario is that for some plugins no actions are shown, which is an acceptable trade-off

Reply:
I suppose there is a potential security concern no matter how you handle it, but it seems like the most sensible approach, and I wonder if there's some way to mitigate the security concern, but I don't have one right away. For what it's worth, it will also produce arbitrary text. Someone could produce a plugin with libelous content for the description, e.g., and Fastlane would end up rendering it on the docs site.

Reply:
Aye that's true, but that was always possible by just giving the plugin an obscene name or something. At least in the action description it's hidden away until a user un-collapses it